brave/src/main/java/brave/propagation/ExtraFieldPropagation.java

```diff
@@ -98,21 +99,131 @@
+  static final Plugin[] NO_CONTEXT_CUSTOMIZERS = new Plugin[0];
+  static final FieldUpdater[] NO_FIELD_UPDATERS = new FieldUpdater[0];
+
+  /**
+   * This allows you to manipulate fields, including dropping or initializing values. This object
+   * will be called for each field added to {@link FactoryBuilder}, in the same order they were,
+   * added and regardless of whether the field is present or not.
+   *
+   * <p>Design note: This is the exact same signature as {@link MutableSpan.TagUpdater}.
+   */
+  public interface FieldUpdater {
+    FieldUpdater NOOP = new FieldUpdater() {
+      @Override public String update(String key, String value) {
+        return value;
+      }
+
+      @Override public String toString() {
+        return "NoopFieldUpdater{}";
+      }
+    };
+
+    /**
+     * Returns the same value, an updated one, or null to drop the field. For example, this could
+     * append to an existing string.
+     *
+     * @see ExtraFieldPropagation#set(String, String)
+     */
+    @Nullable String update(String name, @Nullable String value);
+  }
+
+  /**
+   * Propagation plugins are self-contained and support advanced integration patterns such as
+   * metrics aggregation or sampling overlays.
+   */
+  public static abstract class Plugin {
+    /**
+     * Returns a list of unique lower-case field names used by this plugin.
+     *
+     * <p>The value of this is only read when building {@link ExtraFieldPropagation} and during
+     * {@link #toString()}.
+     */
+    protected abstract List<String> fieldNames();
+
+    /**
+     * This is called once during {@link TraceContext.Extractor#extract(Object)}, allowing you to
+     * decorate the primary trace state with secondary data from extra fields.
+     *
+     * <p>One use case is to override the sampling decision based on an alternate header. If the
+     * decision is intended to be permanent for the rest of the trace, use {@link
+     * TraceContextOrSamplingFlags.Builder#sampled(boolean)}, which affects the primary (usually B3)
+     * headers. Otherwise, you can set {@link TraceContextOrSamplingFlags.Builder#sampledLocal()} to
+     * make an overlaid decision.
+     *
+     * <p>If you are making an overlaid decision, you should either implement {@link
+     * #finishedSpanHandler()} here, or set {@link Tracing.Builder#alwaysReportSpans()} if your
+     * backend can tolerate inconsistent data.
+     *
+     * <p>The resulting field updater will called for each {@link #fieldNames() field name} in the
+     * order they were configured, not in the order headers were received.
+     */
+    protected FieldUpdater extractFieldUpdater(TraceContextOrSamplingFlags.Builder builder) {
+      return FieldUpdater.NOOP;
+    }
+
+    /**
+     * This allows you to customize or prevent extra fields from going to the next host.
+     *
+     * <p>The resulting field updater will called for each {@link #fieldNames() field name} in the
```

Comment:
```suggestion
     * <p>The resulting field updater will be called for each {@link #fieldNames() field name} in the
```

Reply:
otherwise we allocate one throw-away array for each incoming request and each outgoing one, on top of the other overhead added by this feature. This is an attempt to reduce the garbage impact.

Reply:
oops wrong line :)